File: python/quicksort.py
'''Quicksort

  AUTH: Jordan Dodson
  INFO: Implementation of quicksort for instructive purposes.
'''

from helpers import swap, sort_test

def quicksort(arr):
  '''Sort the array by quicksort'''
  _quicksort(arr, 0, len(arr))


def _quicksort(arr, p, r):
  '''Sort the range [p, r) recursively'''

  if r - p > 1:
    q = partition(arr, p, r)
    _quicksort(arr, p, q)
    _quicksort(arr, q+1, r)


def partition(arr, p, r):
  '''Partition the range [p, r) around a pivot'''

  pivot = arr[r-1]
  i = p

  while arr[i] < pivot:
    i += 1

  for j in range(i+1, r-1):
    if arr[j] < pivot:
      swap(arr, i, j)
      i += 1

  # Swap pivot element into place
  swap(arr, i, r-1)

  # Return the partition index
  return i


if __name__ == '__main__':
  sort_test(quicksort, 1000, 1000)

File: js/quicksort.js
function quicksort(arr) {

  _quicksort(arr, 0, arr.length);
}

/**
 * Sort an array in place using the quicksort algorithm
 *
 */
function _quicksort(arr, start, end) {

  if (end <= start) {
    return;
  }

  const middle = partition(arr, start, end);
  _quicksort(arr, start, middle);
  _quicksort(arr, middle + 1, end);

}

/**
 * 
 * Partitions the sub-array [start, end) around a pivot
 */
function partition(arr, start, end) {

  // i < middle => arr[i] < pivot
  let pivotIndex = start;
  const pivot = arr[end - 1];

  while (pivotIndex < end-1 && arr[pivotIndex] < pivot) {
    pivotIndex += 1;
  }
  
  for (let i=pivotIndex+1; i<end-1; i+=1) {
    if (arr[i] < pivot) {
      swap(arr, i, pivotIndex);
      pivotIndex += 1;
    }
  }

  // Swap pivot into place
  swap(arr, pivotIndex, end - 1);

  return pivotIndex;
}

function swap(arr, i, j) {
  const tmp = arr[j];
  arr[j] = arr[i];
  arr[i] = tmp;
}

module.exports.default = quicksort;
